### ISSUE TITLE
Feature request: cross-pool cloning

### ISSUE THREAD

**Issue by outofforest** [openzfs/zfs#12921]
### Describe the feature would like to see added to OpenZFS

Is it possible to implement creating a clone on different pool?

### How will this feature improve OpenZFS?

1. I have a filesystem on a drive. This is a "read-only" storage, nobody modifies it
2. Now I would like to create "writable overlay" on top of it, stored in RAM so writes are as fast as possible

I thought this could work:

1. create RAM device: `modprobe brd rd_nr=1 rd_size=10485760`
2. create pool on this device: `zpool create tmp /dev/ram0`
3. clone filesystem: `zfs clone tank/image@image tmp/image`

But obviously I get error:

```
cannot create 'tmp/image': source and target pools differ
```

### Additional context

I understand that the way `zfs` is designed may make this completely impossible to implement cross-pool cloning. So my first question is if it is possible or not, and second if you consider this use case worth implementing.

**Comment by rincebrain** [comment]
I suppose the question that comes to mind for me is, what's the benefit of implementing something like this versus using Linux's overlayfs, for example?

**Comment by outofforest** [comment]
> I suppose the question that comes to mind for me is, what's the benefit of implementing something like this versus using Linux's overlayfs, for example?

The benefit is exactly the same as in the case of snapshots and existing clones. Both could be implemented using overlayfs, but zfs supports them natively because it pretends to be the best filesystem ever created. Compression and encryption could be done using external software too but zfs supports it for the same reason.

Both snapshots and clones create some kind of diff between "read-only" base and writable overlay. There is no logical reason why those diffs have to be stored on the same pool where base filesystems exist. Storing snapshots as close to the original filesystem as possible (using "@..." suffix) makes sense but clones may already be created everywhere (unless borderline of pool is crossed).

I consider the requirement of storing clone on the same pool to be a limitation which should be fixed unless it's not possible by design.

**Comment by shodanshok** [comment]
@outofforest I really don't see how it can work: different pool have different storage pool allocators (SPAs) and so do not share any blocks. Moreover, what you want should be easily doable via device mapper and/or qcow2 images (via backing files/devices). Am I missing something? What is your specific use case?

**Comment by outofforest** [comment]
@shodanshok My specific use case is described in the description of feature request.
If it's impossible to implement this due to how zfs was designed then please close this request. I will have to deal with my goal using different approach.

**Comment by PaulZ-98** [comment]
Would `zfs send tank/image@image | zfs recv tmp/image` work for your use case?

**Comment by outofforest** [comment]
> Would `zfs send tank/image@image | zfs recv tmp/image` work for your use case?

No, because it will copy all the data. Obviously my goal is to save time and storage space. That's why I would like to use cloning instead.

**Comment by IvanVolosyuk** [comment]
I think it is some kind of data dependency between separate pools. The problem is - the second pool will not be functional without the first, but as the first pool doesn't know about the second one, so it is easily possible to delete the snapshot on the first pool and break the second one. ZFS pool is the top level construct and I think there is no plans to add cross-pool complexity to the code.

**Comment by mskarbek** [comment]
@outofforest simply put "nie ma takiej opcji". You can though use OverlayFS with ZFS. It works well as long as you use ZFS-based file system only as a lower layer of OverlayFS. ZFS in not designed to share data between pools in a way you would like.
```
# zfs create <ZFS_POOL>/lowerdir
# UPPERDIR=$(mktemp -d)
# WORKDIR=$(mktemp -d)
# MOUNTDIR=$(mktemp -d)
# echo "test1" > /<ZFS_POOL>/lowerdir/test
# zfs set readonly=on <ZFS_POOL>/lowerdir
# echo "test2" > /<ZFS_POOL>/lowerdir/test
-bash: /<ZFS_POOL>/lowerdir/test Read-only file system
# mount -t overlay overlay -o lowerdir=/<ZFS_POOL>/lowerdir,upperdir=${UPPERDIR},workdir=${WORKDIR} ${MOUNTDIR}
# cat ${MOUNTDIR}/test                                                                  
test1
# echo "test2" > ${MOUNTDIR}/test
ls ${UPPERDIR}
test
```
